File: C++/ex4/linked_list_reverse_solution.cpp
/*
 * Linked List Reverse (C++ - Exercise 4) - Solution
 * Difficulty: ** (2/5)
 */

#include <iostream>

struct Node {
    int value;
    Node *next;
    Node(int v): value(v), next(nullptr) {}
};

Node* append(Node *head, int value) {
    Node *n = new Node(value);
    if (!head) return n;
    Node *cur = head;
    while (cur->next) cur = cur->next;
    cur->next = n;
    return head;
}

Node* reverse_list(Node *head) {
    Node *prev = nullptr;
    Node *cur = head;
    while (cur) {
        Node *next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }
    return prev;
}

void free_list(Node *head) {
    while (head) {
        Node *tmp = head;
        head = head->next;
        delete tmp;
    }
}

void print_list(Node *head) {
    for (Node *cur = head; cur; cur = cur->next) {
        std::cout << cur->value << (cur->next ? ' ' : '\n');
    }
}

int main() {
    int n;
    if (!(std::cin >> n) || n < 0) {
        std::cerr << "Error: expected non-negative integer n\n";
        return 1;
    }
    Node *head = nullptr;
    for (int i = 0; i < n; ++i) {
        int v;
        if (!(std::cin >> v)) {
            std::cerr << "Error: expected " << n << " integers\n";
            free_list(head);
            return 1;
        }
        head = append(head, v);
    }
    head = reverse_list(head);
    print_list(head);
    free_list(head);
    return 0;
}
/*
 * Linked List Reverse (C++ - Exercise 4) - Solution
 * Difficulty: ** (2/5)
 */

#include <iostream>

struct Node {
    int value;
    Node *next;
    Node(int v): value(v), next(nullptr) {}
};

Node* append(Node *head, int value) {
    Node *n = new Node(value);
    if (!head) return n;
    Node *cur = head;
    while (cur->next) cur = cur->next;
    cur->next = n;
    return head;
}

Node* reverse_list(Node *head) {
    Node *prev = nullptr;
    Node *cur = head;
    while (cur) {
        Node *next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }
    return prev;
}

void free_list(Node *head) {
    while (head) {
        Node *tmp = head;
        head = head->next;
        delete tmp;
    }
}

void print_list(Node *head) {
    for (Node *cur = head; cur; cur = cur->next) {
        std::cout << cur->value << (cur->next ? ' ' : '\n');
    }
}

int main() {
    int n;
    if (!(std::cin >> n) || n < 0) {
        std::cerr << "Error: expected non-negative integer n\n";
        return 1;
    }
    Node *head = nullptr;
    for (int i = 0; i < n; ++i) {
        int v;
        if (!(std::cin >> v)) {
            std::cerr << "Error: expected " << n << " integers\n";
            free_list(head);
            return 1;
        }
        head = append(head, v);
    }
    head = reverse_list(head);
    print_list(head);
    free_list(head);
    return 0;
}

File: C#/ex4/linked_list_reverse_solution.cs
// Solution Ex 4 — Reverse a singly linked list
using System;
using System.Collections.Generic;

class Node {
    public int val;
    public Node? next;
    public Node(int v) { val = v; next = null; }
}

class Program {
    static Node? ReadList(int n) {
        Node? head = null, tail = null;
        int read = 0;
        while (read < n) {
            string? line = Console.ReadLine();
            if (line == null) break;
            var parts = line.Split(new[]{' ','\t'}, StringSplitOptions.RemoveEmptyEntries);
            foreach (var p in parts) {
                if (read >= n) break;
                int v = int.Parse(p);
                var node = new Node(v);
                if (head == null) head = tail = node;
                else { tail!.next = node; tail = node; }
                read++;
            }
        }
        return head;
    }

    static Node? Reverse(Node? head) {
        Node? prev = null;
        var cur = head;
        while (cur != null) {
            var nx = cur.next;
            cur.next = prev;
            prev = cur;
            cur = nx;
        }
        return prev;
    }

    static void Main() {
        string? line = Console.ReadLine();
        if (line == null) return;
        int n = int.Parse(line.Trim());
        var head = ReadList(n);
        head = Reverse(head);
        var outList = new List<int>();
        while (head != null) { outList.Add(head.val); head = head.next; }
        Console.WriteLine(string.Join(' ', outList));
    }
}
